File: core/src/main/java/org/strykeforce/thirdcoast/telemetry/item/AbstractItem.java
package org.strykeforce.thirdcoast.telemetry.item;

import com.squareup.moshi.JsonWriter;
import java.io.IOException;
import java.util.Set;
import org.jetbrains.annotations.NotNull;
import org.strykeforce.thirdcoast.telemetry.grapher.Measure;

/**
 * Abstract base class for Items. This implements {@code Comparable} by comparing the results
 * returned by {@code Item#deviceId()}.
 */
public abstract class AbstractItem implements Item {

  public final String type;
  public final String description;
  public final Set<Measure> measures;

  public AbstractItem(String type, String description, Set<Measure> measures) {
    assert (type != null);
    assert (description != null);
    assert (measures != null);
    this.type = type;
    this.description = description;
    this.measures = measures;
  }

  @Override
  public String type() {
    return type;
  }

  @Override
  public String description() {
    return description;
  }

  @Override
  public Set<Measure> measures() {
    return measures;
  }

  @Override
  public int compareTo(@NotNull Item other) {
    return Integer.compare(deviceId(), other.deviceId());
  }

  @Override
  public void toJson(JsonWriter writer) throws IOException {
    writer.beginObject();
    writer.name("not").value("implemented");
    writer.endObject();
  }

  @Override
  public String toString() {
    return "AbstractItem{"
        + "type='"
        + type
        + '\''
        + ", description='"
        + description
        + '\''
        + ", measures="
        + measures
        + '}';
  }
}

File: telemetry/src/main/kotlin/org/strykeforce/thirdcoast/telemetry/item/AbstractItem.kt
package org.strykeforce.thirdcoast.telemetry.item

import com.squareup.moshi.JsonWriter
import org.strykeforce.thirdcoast.telemetry.grapher.Measure
import java.io.IOException

/**
 * Abstract base class for Items. This implements `Comparable` by comparing the results
 * returned by `Item#deviceId()`.
 * @property type the type of this item.
 * @property description the description of this item.
 * @property measures the measures that are available for this type of item.
 */
abstract class AbstractItem(val type: String, val description: String, val measures: Set<Measure>) : Item {

    override fun type(): String { // FIXME: these need to be properties in interface
        return type
    }

    override fun description(): String {
        return description
    }

    override fun measures(): Set<Measure> {
        return measures
    }

    override fun compareTo(other: Item): Int {
        val result = type.compareTo(other.type())
        return if (result != 0) result else Integer.compare(deviceId(), other.deviceId())
    }

    @Throws(IOException::class)
    override fun toJson(writer: JsonWriter) {
        writer.beginObject()
        writer.name("not").value("implemented")
        writer.endObject()
    }

    override fun toString(): String {
        return ("AbstractItem{"
                + "type='"
                + type
                + '\''.toString()
                + ", description='"
                + description
                + '\''.toString()
                + ", measures="
                + measures
                + '}'.toString())
    }
}
